crates/bevy_animation_rig/src/skinned_mesh.rs

```diff
@@ -0,0 +1,215 @@
+use bevy_asset::{Assets, Handle, HandleUntyped};
+use bevy_ecs::{
+    entity::{Entity, EntityMap, MapEntities, MapEntitiesError},
+    reflect::{ReflectComponent, ReflectMapEntities},
+    system::{Query, Res, ResMut},
+};
+use bevy_math::Mat4;
+use bevy_pbr::render_graph;
+use bevy_reflect::{Reflect, TypeUuid};
+use bevy_render::{
+    pipeline::PipelineDescriptor,
+    render_graph::{base::node, RenderGraph, RenderResourcesNode},
+    renderer::{
+        RenderResource, RenderResourceHints, RenderResourceIterator, RenderResourceType,
+        RenderResources,
+    },
+    shader::{Shader, ShaderStage},
+    texture::Texture,
+};
+use bevy_transform::components::GlobalTransform;
+
+/// Specify RenderPipelines with this handle to render the skinned mesh.
+pub const SKINNED_MESH_PIPELINE_HANDLE: HandleUntyped =
+    HandleUntyped::weak_from_u64(PipelineDescriptor::TYPE_UUID, 0x14db1922328e7fcc);
+
+/// Used to update and bind joint transforms to the skinned mesh render pipeline specified with [`SKINNED_MESH_PIPELINE_HANDLE`].
+///
+/// The length of `joint_entities` and `joint_transforms` should equal to the number of matrices inside [`SkinnedMeshInverseBindposes`].
+///
+/// The content of `joint_transforms` can be modified manually if [`skinned_mesh_update`] system is disabled.
+///
+/// # Example
+/// ```
+/// use bevy_animation_rig::{SkinnedMesh, SKINNED_MESH_PIPELINE_HANDLE};
+/// use bevy_ecs::{entity::Entity, system::Commands};
+/// use bevy_pbr::prelude::PbrBundle;
+/// use bevy_render::pipeline::{RenderPipeline, RenderPipelines};
+///
+/// fn example_system(mut commands: Commands) {
+///     commands.spawn_bundle(PbrBundle {
+///         render_pipelines: RenderPipelines::from_pipelines(
+///             vec![RenderPipeline::new(SKINNED_MESH_PIPELINE_HANDLE.typed())]
+///         ),
+///         ..Default::default()
+///     }).insert(SkinnedMesh::new(
+///         // Refer to [`SkinnedMeshInverseBindposes`] example on how to create inverse bindposes data.
+///         Default::default(),
+///         // Specify joint entities here.
+///         vec![Entity::new(0)]
+///     ));
+/// }
+/// ```
+#[derive(Debug, Default, Clone, Reflect)]
+#[reflect(Component, MapEntities)]
+pub struct SkinnedMesh {
+    pub inverse_bindposes: Handle<SkinnedMeshInverseBindposes>,
+    pub joint_entities: Vec<Entity>,
+    pub joint_transforms: Vec<Mat4>,
```

Comment:
Would it be possible to combine these two vector?
e.g.
```rust
pub joint_entities_transforms: Vec<(Entity, Mat4)>,
```

Reply:
Yes, it is. I just want to make these fields clear. Also, when `mut`ing the transforms, we don't necessarily modify the entities list.

Reply:
Combining vectors seems to make things more convenience. I doubt if the code readability is lowered. What do you think?

Reply:
I'm personally a fan of combining the two vectors, as I think this would improve the code readability.

The main reason is that (afaik) the two vectors are always the same size and `joint_entities[N]` is always used in conjunction with `joint_transforms[N]`.

Separating these into two vectors makes it seems like the two vectors are independent from one another (which they aren't).

Also this prevent having to call `enumerate` when you are iterating one vector in order to index into the second vector.

Reply:
I tried combinding vector by using named struct instead of tuple before. It didn't compile because of thing related to reflection. Using tuple seems to work in this case. I will try it out.

Reply:
My guess is the named struct would also have to derive reflection as well?

Reply:
Sadly, tuple doesn't work either. Using both tuple and struct will panic at runtime if bevy tries to copy component values from glTF scene to app scene.  They result in similar panic error:

```shell
thread 'main' panicked at 'Attempted to push invalid value of type bevy_animation_rig::skinned_mesh::SkinnedMeshJoint.'
```

```shell
thread 'main' panicked at 'Attempted to push invalid value of type (bevy_ecs::entity::Entity, glam::mat4::Mat4).'
```

It happens probably because of `any()` implemented in the derived trait `Reflect` doesn't return `self`. I'm tracking it down.

Reply:
Yea, there's a chance this could be a bug in the reflection code. In which case opening up an issue for this would be nice :)

Reply:
I finally know why. When calling `Reflect::apply(Vec<T>, &Vec<T>)`, it clones vector's elements into `DynamicStruct` by calling `clone_dynamic()` on each element. This results in a panic because it's trying to push `DynamicStruct` into `Vec<T>`.

Reply:
I managed to work around the issue by manually implementing `Reflect` on `SkinnedMeshJoint` 😅

Reply:
This seems like a reflection bug imo.
If you don't mind opening a new issue so we can track this, that'd be awesome :)

Reply:
Okay, I will open an issue later.